pandas/core/tools/datetimes.py

```diff
@@ -305,6 +310,51 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
             except (ValueError, TypeError):
                 raise e
 
+    def _maybe_convert_cache(arg, cache, box, format, name=None, tz=tz):
+        """
+        Try to convert the datetimelike arg using
+        a cache of converted dates.
+
+        Parameters
+        ----------
+        arg : integer, float, string, datetime, list, tuple, 1-d array, Series
+            Datetime argument to convert     
+        cache : boolean
+            If True, try to convert the dates with a cache
+            If False, short circuit and return None
+            Flag whether to cache the converted dates
+        box : boolean
+            If True, return a DatetimeIndex  
+            if False, return an ndarray of values     
+        tz : String or None
```

Comment:
same with these args. The wrapping should be done outside, ``_maybe_cache`` is a raw function that can always just return a ndarray. Then wrap it after its called. otherwise the logic is getting too cumbersome / duplicated.